crates/storage/src/collections/vec/tests.rs

```diff
@@ -422,11 +423,88 @@ fn clear_works_on_filled_vec() {
 
 #[test]
 fn clear_works_on_empty_vec() {
-    let mut vec = vec_from_slice(&[]);
+    let mut vec: StorageVec<()> = vec_from_slice(&[]);
     vec.clear();
     assert!(vec.is_empty());
 }
 
+#[test]
+fn test_binary_search() {
+    let b: StorageVec<i32> = StorageVec::new();
+    assert_eq!(b.binary_search(&5), Err(0));
+
+    let b = vec_from_slice(&[4]);
+    assert_eq!(b.binary_search(&3), Err(0));
+    assert_eq!(b.binary_search(&4), Ok(0));
+    assert_eq!(b.binary_search(&5), Err(1));
+
+    let b = vec_from_slice(&[1, 2, 4, 6, 8, 9]);
+    assert_eq!(b.binary_search(&5), Err(3));
+    assert_eq!(b.binary_search(&6), Ok(3));
+    assert_eq!(b.binary_search(&7), Err(4));
+    assert_eq!(b.binary_search(&8), Ok(4));
+
+    let b = vec_from_slice(&[1, 2, 4, 5, 6, 8]);
+    assert_eq!(b.binary_search(&9), Err(6));
+
+    let b = vec_from_slice(&[1, 2, 4, 6, 7, 8, 9]);
+    assert_eq!(b.binary_search(&6), Ok(3));
+    assert_eq!(b.binary_search(&5), Err(3));
+    assert_eq!(b.binary_search(&8), Ok(5));
+
+    let b = vec_from_slice(&[1, 2, 4, 5, 6, 8, 9]);
+    assert_eq!(b.binary_search(&7), Err(5));
+    assert_eq!(b.binary_search(&0), Err(0));
+
+    let b = vec_from_slice(&[1, 3, 3, 3, 7]);
+    assert_eq!(b.binary_search(&0), Err(0));
+    assert_eq!(b.binary_search(&1), Ok(0));
+    assert_eq!(b.binary_search(&2), Err(1));
+    assert!(match b.binary_search(&3) {
+        Ok(1..=3) => true,
+        _ => false,
+    });
+    assert!(match b.binary_search(&3) {
+        Ok(1..=3) => true,
+        _ => false,
+    });
+    assert_eq!(b.binary_search(&4), Err(4));
+    assert_eq!(b.binary_search(&5), Err(4));
+    assert_eq!(b.binary_search(&6), Err(4));
+    assert_eq!(b.binary_search(&7), Ok(4));
+    assert_eq!(b.binary_search(&8), Err(5));
+
+    let b = vec_from_slice(&[(); u8::MAX as usize]);
+    assert_eq!(b.binary_search(&()), Ok(u8::MAX as u32 / 2));
+}
+
+#[test]
+fn test_binary_search_by_overflow() {
+    let b = vec_from_slice(&[(); u8::MAX as usize]);
+    assert_eq!(
+        b.binary_search_by(|_| Ordering::Equal),
+        Ok(u8::MAX as u32 / 2)
+    );
+    assert_eq!(b.binary_search_by(|_| Ordering::Greater), Err(0));
+    assert_eq!(b.binary_search_by(|_| Ordering::Less), Err(u8::MAX as u32));
+}
+
+#[test]
+// Test implementation specific behavior when finding equivalent elements.
+// It is ok to break this test but when you do a crater run is highly advisable.
```

Comment:
This comment references `crater`, which isn't relevant in our case. We should probably remove that part of the comment.

Reply:
We should leave the first part of the comment ("it is ok to break...") here, it's just the `crater` part that's not relevant